src/Output/ProgressBar.php

```diff
@@ -0,0 +1,300 @@
+        }
+
+        $this->drawProgressBar($current, $label);
+
+        $this->current = $current;
+        $this->label   = $label;
+    }
+
+    /**
+     * Increments the current position we are at and re-writes the progress bar
+     *
+     * @param integer $increment The number of items to increment by
+     */
+    public function advance(int $increment = 1, string $label = '')
+    {
+        $this->current($this->current + $increment, $label);
+    }
+
+    /**
+     * Force the progress bar to redraw every time regardless of whether it has changed or not
+     */
+    public function forceRedraw(bool $force = true): self
+    {
+        $this->forceRedraw = !!$force;
+
+        return $this;
+    }
+
+
+    /**
+     * Update a progress bar using an iterable.
+     *
+     * @param iterable $items Array or any other iterable object
+     * @param callable $callback A handler to run on each item
+     */
+    public function each($items, callable $callback = null)
+    {
+        if ($items instanceof \Traversable) {
+            $items = iterator_to_array($items);
+        }
+
+        $total = count($items);
+        if (!$total) {
+            return;
+        }
+        $this->total($total);
+
+        foreach ($items as $key => $item) {
+            if ($callback) {
+                $label = $callback($item, $key);
+            } else {
+                $label = null;
+            }
+
+            $this->advance(1, $label);
+        }
+    }
+
+
+    /**
+     * Draw the progress bar, if necessary
+     */
+    protected function drawProgressBar(int $current, string $label)
+    {
+        $percentage = $this->percentageFormatted($current / $this->total);
+
+        if ($this->shouldRedraw($percentage, $label)) {
+            $progress_bar = $this->getProgressBar($current, $label);
+            $this->writer->write($progress_bar, true);
+        }
+
+        $this->currentPercentage = $percentage;
+    }
+
+    /**
+     * Build the progress bar str and return it
+     */
+    protected function getProgressBar(int $current, string $label): string
+    {
+        if ($this->firstLine) {
+            // Drop down a line, we are about to
+            // re-write this line for the progress bar
+            $this->writer->write('');
+            $this->firstLine = false;
+        }
+
+        // Move the cursor up and clear it to the end
+        $line_count = $this->hasLabelLine ? 2 : 1;
+
+        $progress_bar  = $this->cursor->up($line_count);
+        $progress_bar .= "\r";
+        $progress_bar .= $this->cursor->eraseLine();
+        $progress_bar .= $this->getProgressBarStr($current, $label);
+
+        // If this line has a label then set that this progress bar has a label line
+        if (strlen($label) > 0) {
+            $this->hasLabelLine = true;
+        }
+
+        return $progress_bar;
+    }
+
+    /**
+     * Get the progress bar string, basically:
+     * =============>             50% label
```

Comment:
pretty nice. it would be even better if we can choose progressbar char from a set eg `=`, `■`, `▧`, `▨`, `▦`, `▩`  etc with default `=` :) 
which seems as simple as swapping the `options['loader']`

Reply:
that's the idea of the options property. it is planned to put a setter which will modify the pointer, the loader and the color of the progress bar.
we can do something like 
`$progress->option('loader', '*');` or 
`$progress->option([
    'loader' => '*',
    'pointer' => '>>'
]);`